### DISCUSSION TITLE
Summary of Metadata of items - Sum of ItemXYZ.Value

### DISCUSSION THREAD

**Discussion by MeikTranel** [dotnet/msbuild#7667] [Q&A] [answered]
Is there any way to resolve the sum of all `Value`s here?

```xml
<ItemGroup>
  <ItemXYZ Include="A" Value="5" />
  <ItemXYZ Include="B" Value="2" />
  <ItemXYZ Include="C" Value="3" />
</ItemGroup>
```

**Comment by jrdodds** [comment, marked as answer]
Are you asking how to sum the numbers, i.e. 5 + 2 + 3 = 10?

Here is a solution that uses an inline task.

```XML
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <ItemXYZ Include="A" Value="5" />
    <ItemXYZ Include="B" Value="2" />
    <ItemXYZ Include="C">
      <Value>3</Value>
    </ItemXYZ>
  </ItemGroup>

  <Target Name="Main">
    <Sum Addends="@(ItemXYZ->Metadata('Value'))">
      <Output TaskParameter="Total" PropertyName="Sum" />
    </Sum>
    <Message Text="Sum = $(Sum)" />
  </Target>

  <UsingTask TaskName="Sum" TaskFactory="RoslynCodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll">
    <ParameterGroup>
      <Addends ParameterType="Microsoft.Build.Framework.ITaskItem[]" Required="true" />
      <Total ParameterType="System.Int32" Output="true" />
    </ParameterGroup>
    <Task>
      <Code Type="Fragment" Language="cs">
        <![CDATA[
            Total = 0;
            foreach (var addend in Addends)
            {
                if (int.TryParse(addend.ItemSpec, out int result))
                {
                    Total += result;
                }
            }
        ]]>
      </Code>
    </Task>
  </UsingTask>
</Project>
```

The `Sum` task takes an ItemGroup and returns the sum of the items that can be parsed to integers.

The `Sum` task just generically uses the ItemSpec. Your example ItemGroup has a metadata value named 'Value'. The Metadata function is used to create a new ItemGroup that is the values of the 'Value' metadata, `@(ItemXYZ->Metadata('Value'))`.

**Reply by rainersigwald** [reply]
Yeah, IMO this is pretty solidly in the "even if you can figure out a way to do it in MSBuild XML, everyone will be happier if you write a task" category.

**Reply by jrdodds** [reply]
Yeah, I rather quickly reached the conclusion to write a task.

**Reply by MeikTranel** [reply]
Yeaaaaaaaah i was hoping for a shorter syntax but the task works of course. my use case is that i have a `ToolTask` derived testrunner that returns report items with a shorthand summary of all outcomes inside the metadata of that report. I send that metadata as telemetry to an elk backend and i wanted to send another item summing the entire testrun for all test suites - problem is that i have like NUnit.xml amounts of outcomes which means with this solution i would have like 30 lines of xml just to generate the properties.

Thanks for lending me some brain cycles tho.
